File: LeetCode/Search in Rotated Sorted Array II.cpp
class Solution {
public:
    bool search(int A[], int n, int target) {
        int left = 0, right = n-1, mid;

        while (left <= right){
            while(left <= right){
                if (left + 1 < n && A[left] == A[left+1]) left++;
                else if (right - 1 >= 0 && A[right] == A[right-1]) right--;
                else break;
            }
            mid = (right - left) / 2 + left;

            if (target == A[mid])
                return true;
            else if ((A[left] <= A[mid] && target < A[mid] && target >= A[left]) ||
                (A[left] > A[mid] && !(target > A[mid] && target <= A[right])))
                right = mid - 1;
            else if ((A[left] > A[mid] && target > A[mid] && target <= A[right]) ||
                (A[left] <= A[mid] && !(target < A[mid] && target >= A[left])))
                left = mid + 1;
        }

        return false;
    }
};

File: LeetCode/Search in Rotated Sorted Array II.py
class Solution:
	# @param A a list of integers
	# @param target an integer
	# @return a boolean
	def search(self, A, target):
		# find the split pos
		low = 0
		high = len(A) - 1
		mid = -1

		while low <= high:
			while high-1 >= 0 and A[high] == A[high-1]:
				high -= 1
			mid = (low + high) / 2
			# print (low, high)
			if A[mid] > A[low] and A[mid] < A[high]:
				break
			elif A[mid] > A[low]:
				low = mid
			elif A[mid] < A[high]:
				high = mid
			else:
				break

		while mid+1 < len(A) and A[mid+1] >= A[mid]:
			mid += 1

		high = mid
		if mid + 1 < len(A):
			low = mid + 1
		else: low = -1

		if target <= A[high] and target >= A[0]:
			low = 0
		elif low!= -1 and target >= A[low] and target <= A[len(A)-1]:
			high = len(A)-1
		else:
			return False

		# binary search
		while low <= high:
			mid = (low + high) / 2
			if A[mid] < target:
				low = mid + 1
			elif A[mid] > target:
				high = mid - 1
			else:
				return True

		return False

if __name__ == '__main__':
	s = Solution()
	A = [1, 2, 3, 5, 5, 5, 0]
	print s.search([1, 2, 3, 5, 5, 5, 0, 1], 5)
	print s.search([5, 5, 5], 3)
	print s.search([5, 5, 1, 1, 5, 5], 5)
	print s.search([11], 5)
	print s.search([3,0,0,1,2,2], 3)